pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package rafttrace
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/raft"
+	"github.com/cockroachdb/cockroach/pkg/raft/raftpb"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing/tracingpb"
+)
+
+type traceValue struct {
+	// ctx is a trace specific context used to log events on this trace.
+	ctx context.Context
+	// baseCtx is the underlying proposal buffer which we additionally log to.
+	baseCtx withContext
+	// cleanup will retrieve all events that were logged in this trace.
+	cleanup func() tracingpb.Recording
+}
+
+// RaftTracer is a utility to trace raft messages. Note that it may erroneously
+// include incorrect messages since it does not consider the term, however it is
+// easier to filter out extra messages that deal with missing messages.
+type RaftTracer struct {
+	// If this is too slow consider a bloom filter to see if this event is
+	// possibly traced. This map should normally be very small.
+	m syncutil.Map[uint64, traceValue]
```

Comment:
Under assumption that we can't use `Map`, and need some data structure that understands the order of entries in the log, we likely need a mutex. Then we'll need a fast-path to avoid locking this mutex on every message.

Could be something like:

```golang
type RaftTracer struct {
	active atomic.Bool
	mu syncutil.Mutex
	traced []tracedEntry
}

func (r *RaftTracer) MaybeTrace(m raftpb.Message) {
	if !r.active.Load() {
		return
	}
	r.mu.Lock()
	// Intersect. Find a single entry affected by the message,
	// or a range of entries. We should be able to do it in O(1)
	// or O(log N) depending on the data structures / messages.
	// maybe call r.unregisterLocked(index) for relevant events.
	r.mu.Unlock()
	// do the logging
}

func (r *RaftTracer) Register(traceID, ...) {
	r.mu.Lock()
	r.active.Store(true)
	// enqueue
	r.mu.Unlock()
}

func (r *RaftTracer) unregisterLocked(index) {
	// dequeue up to index
	if len(r.traced) == 0 {
		r.active.Store(false)
	}
}
```

Reply:
I was looking for some type of lock free structure, but agreed if we decide to maintain the order we would need this.

Reply:
I think this should be OK now - let me know if you have any concerns. Based on the other comments I think we are OK.